common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright 2015-2016 IBM Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.database
+
+import java.util.concurrent.atomic.AtomicReference
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+import scala.concurrent.Promise
+import scala.concurrent.duration.DurationInt
+import scala.language.implicitConversions
+import scala.util.Failure
+import scala.util.Success
+
+import spray.caching.Cache
+import spray.caching.LruCache
+import spray.caching.ValueMagnet.fromAny
+import whisk.common.Logging
+import whisk.common.LoggingMarkers
+import whisk.common.TransactionId
+
+/**
+ * A cache that allows multiple readers, bit only a single writer, at
+ * a time. It will make a best effort attempt to coalesce reads, but
+ * does not guarantee that all overlapping reads will be coalesced.
+ *
+ * The cache operates by bracketing all reads and writes. A read
+ * imposes a lightweight read lock, by inserting an entry into the
+ * cache with State.ReadInProgress. A write does the same, with
+ * State.WriteInProgress.
+ *
+ * On read or write completion, the value transitions to
+ * State.Cached.
+ *
+ * State.Initial is represented implicitly by absence from the cache.
+ *
+ * The handshake for cache entry state transition is:
+ * 1. if entry is in an agreeable state, proceed
+ *    where agreeable for reads is Initial, ReadInProgress, or Cached
+ *                    and the read proceeds by ensuring the entry is State.ReadInProgress
+ *      and agreeable for writes is Initial or Cached
+ *                    and the write proceeds by ensuring the entry is State.WriteInProgress
+ *
+ * 2. if entry is not in an agreeable state, fail fast with Future.failed(new ConcurrentUpdateException)
+ *
+ * 3. to swap in the new state to an existing entry, we use an AtomicReference.compareAndSet
+ *
+ * 4. after adding a new entry, we double check that the state of the actually-added entry
+ *    is agreeable with the desired action (as per item 1)
+ *
+ */
+trait MultipleReadersSingleWriterCache[W, Winfo] {
+  val DEBUG = true
```

Comment:
i think this would be fine. my only word of caution is the DEBUG stuff is in there really just for development, and is fairly liberally placed. is the scala compiler smart enough to excise logger.debug calls if the log level is set higher?